src/utils.rs

```diff
@@ -90,6 +91,14 @@ pub fn add_le(x: &mut [u8], y: &[u8]) -> Result<(), ()> {
     }
 }
 
+pub(crate) fn memzero_value<T>(mut v: T) {
```

Comment:
Maybe this should indicate that it is actually dropping/forgetting value, like `memzero_and_forget_value`? (And maybe `debug_assert!(!needs_drop(...));`?)

Reply:
Or `drop_memzero` for consistency with `Digest`?

Reply:
Actually, one could avoid changing the API of `Digest` and `State` by just providing a public `forget_memzero` function that could just be used like `forget_memzero(digest);` which is not more complicated than `digest.drop_memzero();` and aligns nicely with the standard libraries `mem::forget` and `mem::drop` functions.

Reply:
This function is unsafe. Being able to zero out any value will cause undefined behavior. Also it won’t work for any type that keeps its values anywhere else than in the struct, causing values to stay in memory, memory leaks and so on.

Reply:
> This function is unsafe. Being able to zero out any value will cause undefined behavior. 

I though so too at first, but having taken ownership of the value and `forget`ting it before zeroing out the memory should actually be safe as nothing is touching the memory anymore.

> Also it won’t work for any type that keeps its values anywhere else than in the struct, causing values to stay in memory, memory leaks and so on.

Indeed, but failing to clear everything and leaking allocations is not a memory safety issue. (`forget` itself is not unsafe either). But a `T: Copy` bound would probably be a good thing if it is intended to improve security.

One other issue issue is taking ownership could imply `memcpy` behind the scenes, but AFAIK this affects the other zero-on-drop types as well.

Reply:
The leaking is actually not strictly necessary either as `ptr::drop_in_place` could be used instead of `mem::forget`.

Reply:
Leaking isn’t unsafe, it’s just bad. Overwriting memory with bit patterns the type can’t contain is undefined behavior, no matter if the value is dropped afterwards or not, and no matter if the type is copy or not. See the whole raison d’etre for MaybeUninit.

Reply:
> Overwriting memory with bit patterns the type can’t contain is undefined behavior, no matter if the value is dropped afterwards or not, and no matter if the type is copy or not. See the whole raison d’etre for MaybeUninit.

Interpreting an arbitrary a bit pattern as `T`, e.g. `bool`, is UB. But zeroing memory at an arbitrary address if there is no more `T` is not. `MaybeUninit` makes it easier to work with uninitialized memory and avoid the pitfalls of `mem::uninitialized`, but there is nothing preventing one from doing all of the same using only raw pointers.

In this case, this is exactly like deallocting a heap-allocated value, e.g. [`impl Drop for Rc`](https://doc.rust-lang.org/src/alloc/rc.rs.html#1065-1110) (`Box` is special at it is currently impemented in the compiler), i.e. the value is destroy using `ptr::drop_in_place` and then it is deallocated using `Global.dealloc`. Nothing prevents us from zero'ing out the memory in between these steps as there is no more `T` the invariants of which we need to uphold.

Reply:
(Note that there is no `&mut T` or even `*mut T` in play after `drop`ping, only `*mut u8` resp. `&mut [u8]`.)

Reply:
I think `MaybeUnint` would actually make this nicer to write, e.g.
```
fn drop_memzero<T: Copy>(mut v: T) {
  let v = mem::MaybeUninit::new(v);
  unsafe {
        ptr::drop_in_place(v.as_mut_ptr());
        ffi::sodium_memzero(v.as_mut_ptr() as _, mem::size_of::<T>());
  }
}
```

Reply:
`forget_memzero(state)` isn't possible anyway since `finalize` consumes the state

Reply:
> forget_memzero(state) isn't possible anyway since finalize consumes the state

Right, I did not see that. But then, the actual finalize logic should probably be factored out into a `&mut self` method to avoid repeating it. The question whether `drop_memzero` resp. `memzero_value` are safe still needs to answered tough.

Reply:
State is invalid after finalizing. Original libsodium documentation clearly states that state can't be used again after finalizing, so it needs to be consumed.

Reply:
> Original libsodium documentation clearly states that state struct can't be used again after finalizing, so it needs to be consumed.

Yes, and the public API should enforce this. That does not mean that an internal helper method could not take `&mut self` to finalize, it should just not be public.

Reply:
Yeah, it makes sense. I'll push somethig on that matter soon